### PR TITLE
stub: Fix splash alpha blending

### PR DESCRIPTION
**Pull request by medhefgo**

### CODE REVIEW

**Comment by medhefgo** [comment]
While giving https://github.com/systemd/systemd/pull/25180 a spin I learned that our BMP parser does not handle alpha correctly. Now I know way more about BMPs than I ever cared for.

File: src/boot/efi/splash.c [hunk 1]
```diff
@@ -92,38 +87,26 @@ static EFI_STATUS bmp_parse_header(
...
                         map_count = dib->colors_used;
-                else {
-                        switch (dib->depth) {
-                        case 1:
-                        case 4:
-                        case 8:
-                                map_count = 1 << dib->depth;
-                                break;
+                else if (IN_SET(dib->depth, 1, 4, 8))
+                        map_count = 1 << dib->depth;
```
Comment (bluca) [inline src/boot/efi/splash.c:107]: shouldn't there be an `else map_count = 0` here?

**Review by bluca — APPROVED** [review]
I'm not even going to pretend to understand what's happening here

File: src/boot/efi/splash.c (hunk 1 above)
Comment (medhefgo) [reply src/boot/efi/splash.c:107]: That's been moved to the variable initialization.

Comment (bluca) [reply src/boot/efi/splash.c:107]: duh!

**Comment by medhefgo** [comment]
> I'm not even going to pretend to understand what's happening here

Let me help you out there. :P

Basically, instead of having pixels be stored in `AARRGGBB`, they can be in any form like `BBGGRRAA` for example. In fact they could be in `ARRRGBBB` form. To figure out which bits belong to which color channel you use the mask. The shift then moves the bits to bit offset 0, and the scale widens the value to full 8bits (so a 1bit in alpha channel becomes a proper 0xFF alpha value).

File: src/boot/efi/splash.c [hunk 2]
```diff
@@ -135,28 +135,51 @@ static EFI_STATUS bmp_parse_header(
         return EFI_SUCCESS;
 }
 
-static void pixel_blend(uint32_t *dst, const uint32_t source) {
-        uint32_t alpha, src, src_rb, src_g, dst_rb, dst_g, rb, g;
+enum Channels { R, G, B, A, _CHANNELS_MAX };
```
Comment (poettering) [inline src/boot/efi/splash.c:121]: hmm, you name this enum "Channels", but never actually make use of the type name. Might as well make this anonymous...

I find "R" a *very* short identifier. Maybe at least call this CHANNEL_R? and similar for the others...

**Pull request merged**